sound/soc/sof/ipc4-pcm.c

```diff
@@ -59,19 +59,120 @@ int sof_ipc4_set_pipeline_state(struct snd_sof_dev *sdev, u32 id, u32 state)
 }
 EXPORT_SYMBOL(sof_ipc4_set_pipeline_state);
 
+static void
+sof_ipc4_add_pipeline_to_trigger_list(struct snd_sof_dev *sdev, int state,
+				      struct snd_sof_pipeline *spipe,
+				      struct ipc4_pipeline_set_state_data *trigger_list)
+{
+	struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+	if (pipeline->skip_during_fe_trigger)
+		return;
+
+	switch (state) {
+	case SOF_IPC4_PIPE_RUNNING:
+		/*
+		 * Trigger pipeline if all PCMs containing it are paused or if it is RUNNING
+		 * for the first time
+		 */
+		if (spipe->started_count == spipe->paused_count)
+			trigger_list->pipeline_ids[trigger_list->count++] =
+				pipe_widget->instance_id;
+		break;
+	case SOF_IPC4_PIPE_RESET:
+		/* RESET if the pipeline is neither running nor paused */
+		if (!spipe->started_count && !spipe->paused_count)
+			trigger_list->pipeline_ids[trigger_list->count++] =
+				pipe_widget->instance_id;
+		break;
+	case SOF_IPC4_PIPE_PAUSED:
+		/* Pause the pipeline only when its started_count is 1 more than paused_count */
+		if (spipe->paused_count == (spipe->started_count - 1))
+			trigger_list->pipeline_ids[trigger_list->count++] =
+				pipe_widget->instance_id;
+		break;
+	default:
+		break;
+	}
+}
+
+static void
+sof_ipc4_update_pipeline_state(struct snd_sof_dev *sdev, int state, int cmd,
+			       struct snd_sof_pipeline *spipe,
+			       struct ipc4_pipeline_set_state_data *trigger_list)
+{
+	struct snd_sof_widget *pipe_widget = spipe->pipe_widget;
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+	int i;
+
+	if (pipeline->skip_during_fe_trigger)
+		return;
+
+	/* set state for pipeline if it was just triggered */
+	for (i = 0; i < trigger_list->count; i++) {
+		if (trigger_list->pipeline_ids[i] == pipe_widget->instance_id) {
+			pipeline->state = state;
+			break;
+		}
+	}
+
+	switch (state) {
+	case SOF_IPC4_PIPE_PAUSED:
+		switch (cmd) {
+		case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+			/*
+			 * increment paused_count if the PAUSED is the final state during
+			 * the PAUSE trigger
+			 */
+			spipe->paused_count++;
+			break;
+		case SNDRV_PCM_TRIGGER_STOP:
+		case SNDRV_PCM_TRIGGER_SUSPEND:
+			/*
+			 * decrement started_count if PAUSED is the final state during the
+			 * STOP trigger
+			 */
+			spipe->started_count--;
+			break;
+		default:
+			break;
+		}
+		break;
+	case SOF_IPC4_PIPE_RUNNING:
+		switch (cmd) {
+		case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+			/* decrement paused_count for RELEASE */
+			spipe->paused_count--;
+			break;
+		case SNDRV_PCM_TRIGGER_START:
+		case SNDRV_PCM_TRIGGER_RESUME:
+			/* increment started_count for START/RESUME */
+			spipe->started_count++;
+			break;
+		default:
```

Comment:
I will need paper and pan to visualize this state management...
I think I don't get this. :(

Initially `spipe->paused_count` = `spipe->started_count` = 0
Then we start a  stream.
The first step is:
`sof_ipc4_update_pipeline_state(sdev, SOF_IPC4_PIPE_PAUSED, SNDRV_PCM_TRIGGER_START, ,)`
`spipe->paused_count`, `spipe->started_count` remains 0
Will the pipeline going to be set to PAUSED out from the RESET (is there a need to this at all)?

Next:
`sof_ipc4_update_pipeline_state(sdev, SOF_IPC4_PIPE_RUNNING, SNDRV_PCM_TRIGGER_START, ,)`
`spipe->paused_count` = 0, `spipe->started_count` = 1

Right, this kind of make sense.

Can we receive TRIGGER_STOP while the stream is paused? That might leave the paused_count unbalanced?

Reply:
no PAUSE is always done with a release followed by a stop.

Reply:
> I will need paper and pan to visualize this state management...

Please do use a pen and paper, thats how I visualized this for implemeting it

Reply:
> no stop is always done with a release followed by a stop if the pcm is paused during stop

Humm, it's perfectly legit to stop a paused stream in ALSA.

Reply:
yes but when you stop a paused stream, it always goes release and then stop

Reply:
in DPCM, we go directly from PAUSED to STOP states without ever doing a PAUSE_RELEASE

````
case SNDRV_PCM_TRIGGER_STOP:
			if ((be->dpcm[stream].state != SND_SOC_DPCM_STATE_START) &&
			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED))
				goto next;
````

Can you point me to the transition where we are going to do a PAUSE_RELEASE and STOP?

Reply:
I dont know about where this code is @plbossart . But you can try pausing a stream and then killing it without releasing it and you'll see the transition with 2 triggers, first one for pause_release and then the next one for stop

Reply:
@ranj063 @ujfalusi I am pretty sure an FE can go from PAUSE_PUSH to STOP. I did add this comment in one of the soc-pcm changes.

"
To track the special case of an FE going from PAUSE_PUSH to STOP, we
    add a state variable for each FE context. This 'fe_pause' boolean is
    set on PAUSE_PUSH and cleared on either PAUSE_RELEASE and STOP
    triggers.
"